sources/host-ctr/pidnamespaces/pid_namespaces.go

```diff
@@ -0,0 +1,169 @@
+func AddPidNsToManifest(pidNs int64) (err error) {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	ok, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil && !ok {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	defer func() {
+		if err = fileLock.Unlock(); err != nil {
+			err = errors.Wrap(err, "failed to release lock for writing to PID NS manifest")
+		}
+	}()
+
+	// Add task PID NS to manifest
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
+		return errors.Wrap(err, "failed to read PID NS manifest file")
+	}
+	tmpPidNsList := &HostCtrPidNs{}
+	hostCtrPidNsList := &HostCtrPidNs{}
+	if len(raw) != 0 {
+		err = json.Unmarshal(raw, &tmpPidNsList)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal PID NS manifest")
+		}
+		for _, ele := range tmpPidNsList.PidNamespaces {
+			// If it already exists in the manifest list for whatever reason, we can just return without doing anything.
+			if ele == pidNs {
+				return nil
+			}
+		}
+	}
+	hostCtrPidNsList.PidNamespaces = append(tmpPidNsList.PidNamespaces, pidNs)
+	if err = writeManifest(hostCtrPidNsList, manifest); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemovePidNsFromManifest takes a PID's namespace and removes it from the internal manifest
+func RemovePidNsFromManifest(pidNs int64) (err error) {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	ok, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil && !ok {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	defer func() {
+		if err = fileLock.Unlock(); err != nil {
+			err = errors.Wrap(err, "failed to release lock for writing to PID NS manifest")
+		}
+	}()
+
+	// Remove task PID NS from manifest
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
+		return errors.Wrap(err, "failed to read PID NS manifest file")
+	}
+	hostCtrPidNsList := &HostCtrPidNs{}
+	if len(raw) != 0 {
+		err = json.Unmarshal(raw, &hostCtrPidNsList)
```

Comment:
Maybe factor out a helper that just turns a file path (`manifestPath`) into a `HostCtrPidNs{}`.